### PR TITLE
fix(modal): leaked CSS

### PR DESCRIPTION
**Pull request by lkritsimas**
**Describe pull-request**  
Before this change, third-party Stencil applications using Tegel together with @stencil/sass and shadow DOM would get the Tegel modal component styling applied to them, and they'd end up looking pretty strange. 

This happened because the modals `:host` pseudo-class got included in the `tegel.css` dist file *(probably by accident)*.

The classes `.header` and `.body` were also leaking and could cause conflicts with user styling.

**To prevent this from happening again in the future**, it might be a good idea to analyze the dist stylesheet. 
You could perhaps use a second Stylelint config and lint the dist stylesheet after a build to ensure things like `:host` stay out of this file.
https://stylelint.io/user-guide/rules/selector-pseudo-class-disallowed-list/

You probably also want to ensure that all classes in the dist stylesheet always begin with the prefix `.tds-`.

**How to test**

TL;DR:
1. Run storybook.
2. Verify that the modal component looks like it should.
3. Verify that everything else looks like it should.
4. Verify that dist stylesheet does not contain `:host`, `.header` or `.body`.
5. Cry, because you need to follow all of the instructions below in order to actually test the shadow DOM behavior.

*The instructions below are for testing on your local machine.*

1. `git clone https://github.com/lkritsimas/tegel.git`
2. `git checkout fix/modal-component-leaks-css-and-breaks-shadow-dom-in-host-app`
3. `cd tegel`
4. `npm i`
5. `npm run build`
6. `cd ..`
7. Create a new Stencil project `npm init stencil`
8. Select `component` in the first step of Stencil setup.
9. Enter `shadow-dom-test` as project name and complete setup.
10. `cd shadow-dom-test`
11. `npm i @scania/tegel@^1 @stencil/sass` 
12. Create the file `src/global.scss` with the following content *(first step to enable Tegel in the shadow DOM)*:
    ```scss
    @import '~@scania/tegel/dist/tegel/tegel.css'; 
    ```
13. Edit `stencil.config.ts`, add the following code in `config`:
    ```ts
    import { sass } from '@stencil/sass';
    ```
    ```ts
      plugins: [
        sass({
          // Bake the Tegel stylesheet into all components that are using .scss stylesheets, enabling Tegel in shadow DOM.
          // For testing purposes only, you don't want to do this in a real project since it will bloat the build.
          injectGlobalPaths: [
            'src/global.scss'
          ]
        })
      ],
    ```
    Replace the `www` `outputTarget` with the following code to access the Tegel stylesheet in the light DOM on the dev server.
    ```ts
        {
          type: 'www',
          serviceWorker: null, // disable service workers
          copy: [
            {
              src: '../node_modules/@scania/tegel/dist/tegel/tegel.css',
              dest: 'tegel.css'
            }
          ]
        },
    ```
14. Edit `src/index.html` and add the following line to `<head>` to enable Tegel in the "light DOM". Otherwise we can't access things like `:root` styling properly from the shadow DOM. The Tegel web components also depend on this being loaded outside the shadow DOM in order to render with proper styling.
    ```html
    <link rel="stylesheet" href="tegel.css" />
    ```
15. Open `src/components/my-component.tsx` and replace the contents of this default Stencil component with:
    ```tsx
    import { Component, State, h } from '@stencil/core';
    import '@scania/tegel';

    @Component({
      tag: 'my-component',
      styleUrl: 'my-component.scss',
      shadow: true
    })
    export class MyComponent {
      @State() modalTriggerRef?: HTMLElement;

      render() {
        return (
          <div>
            {/*
              This blue headline will only render correctly if the Tegel CSS classes are passed to the shadow DOM.
              Disable `injectGlobalPaths` and restart dev server. Then you'll see that the Tegel web components will render, but "standalone" CSS classes will not.
            */}
            <p class="tds-headline-02 tds-text-blue-500">Main component</p>
            <tds-button ref={(ref) => this.modalTriggerRef = ref} text="Open modal" />
            <tds-modal
              header="Test"
              referenceEl={this.modalTriggerRef}
              size="lg"
              actions-position="static"
            >
              <span slot="body">
                The steps fell lightly and oddly, with a certain swing, for all they went so slowly; it was
                different indeed from the heavy creaking tread of Henry Jekyll. Utterson sighed. “Is there never
                anything else?” he asked.
              </span>
              <span slot="actions">
                <tds-button data-dismiss-modal size="md" text="Sure thing!" />
              </span>
            </tds-modal>
            <my-child-component />
          </div>
        );
      }
    }
    ```
16. Rename `src/components/my-component.css` to `src/components/my-component.scss`
17. Create a new component called `my-child-component.tsx` with the following content:
    ```tsx
    import { Component, h } from '@stencil/core';

    @Component({
      tag: 'my-child-component',
      styleUrl: 'my-child-component.scss',
      shadow: true
    })
    export class MyChildComponent {
      render() {
        return (
          <div>
            {/*
              This red headline will only render correctly if the Tegel CSS classes are passed to the shadow DOM.
              Disable `injectGlobalPaths` and restart dev server. Then you'll see that the Tegel web components will render, but "standalone" CSS classes will not.
            */}
            <p class="tds-headline-02 tds-text-red-500">Child component</p>
            <tds-accordion>
              <tds-accordion-item header="First item">
                This is the panel, which contains associated information with the header. Usually it contains
                text, set in the same size as the header. Lorem ipsum doler sit amet.
              </tds-accordion-item>
              <tds-accordion-item expanded>
                <div slot="header">Second item</div>
                This is the panel, which contains associated information with the header. Usually it contains
                text, set in the same size as the header. Lorem ipsum dolor sit amet, consectetur adipiscing
                elit. Duis laoreet vestibulum fermentum.
              </tds-accordion-item>
            </tds-accordion>
          </div>
        );
      }
    }
    ```
18. Create a new SCSS stylesheet called `my-child-component.scss` next to the component. The contents of this file doesn't matter, but it must exist in order for `injectGlobalPaths` to work.
19. Run the dev server `npm run start`
20. **Notice how the components styling looks pretty weird.**
21. Stop the dev server.
22. Install the local Tegel package that we built earlier `npm i "../tegel/packages/core"`
23. Run the dev server again `npm run start`
24. **Notice how the styling now looks as expected.**
25. Award yourself with a nice break for spending the time following these steps.

**Screenshots**  
Before:
tegel modal - bad

After:
tegel modal - good

### CODE REVIEW

**Comment by mJarsater** [comment]
Great stuff!

Testing this right now

Really useful to have you guys use Stencil to point these stuff out 🎈 Thinking if this could have an impact in any other application right now? Could it be a breaking change for other teams? If so, we would need to wait with this until 2.0.

File: packages/core/src/global/global.scss [hunk 1]
```diff
@@ -41,7 +41,7 @@
 @import '../components/header/header-vars';
 @import '../components/link/link-vars';
 @import '../components/message/message-vars';
-@import '../components/modal/modal';
+@import '../components/modal/modal-vars';
```
Comment (mJarsater) [inline packages/core/src/global/global.scss:44]: 🧐

**Review by mJarsater — APPROVED** [review]
Steak sauce 🥩

**Review by timrombergjakobsson — APPROVED** [review]

**Comment by lkritsimas** [comment]
If it's a breaking change or not is a good question! It could be, if a user has decided to use the styling of any of these classes that made its way into the main css file on a normal HTML element.

All of the CSS below was accessible for anyone to use, but since there was no documentation for these classes I guess you can only hope no one used them 😅 

If that sounds too risky then we could instead just move the CSS with the `:host` class and leave the `import` as is for now and change that in 2.0. Then maybe a temporary separate .scss file can be created to hold the `:host` styling, and that is what the component will use.

```scss
.tds-modal {
  box-sizing: border-box;
  background-color: var(--tds-modal-background);
  margin: auto;
  position: relative;
  border-radius: 4px;
  padding: 16px;
  max-height: 85vh;
  overflow-y: auto;
}

.tds-modal * {
  box-sizing: border-box
}

.tds-modal::-webkit-scrollbar {
  width: 5px;
  background-color: var(--tds-grey-300);
  border-radius: 0 1em 1em 0
}

.tds-modal::-webkit-scrollbar-track {
  background-color: var(--tds-grey-300);
  border-radius: 0 1em 1em 0
}

.tds-modal::-webkit-scrollbar-thumb {
  background-color: var(--tds-grey-500);
  border-radius: 0 1em 1em 0
}

.tds-modal__actions-sticky {
  overflow: hidden
}

.tds-modal__actions-sticky .body {
  font: var(--tds-body-01);
  letter-spacing: var(--tds-body-01-ls);
  padding-bottom: 96px;
  margin: 0 -16px 0 0;
  max-height: calc(85vh - 36px);
  overflow-y: auto
}

.tds-modal__actions-sticky slot[name=actions] {
  position: absolute;
  bottom: -1px;
  left: 0;
  right: 0;
  background-color: var(--tds-modal-background);
  padding: var(--tds-spacing-element-16);
  display: flex;
  gap: 16px
}

.tds-modal__actions-static slot[name=actions] {
  background-color: var(--tds-modal-background);
  display: flex;
  gap: 16px
}

@media (min-width: 320px) {
  .tds-modal-xs {
    width: 100%
  }

  .tds-modal-sm {
    width: 100%
  }

  .tds-modal-md {
    width: 100%
  }

  .tds-modal-lg {
    width: 100%
  }
}

@media (min-width: 672px) {
  .tds-modal-xs {
    width: 75%
  }

  .tds-modal-sm {
    width: 62.5%
  }

  .tds-modal-md {
    width: 75%
  }

  .tds-modal-lg {
    width: 100%
  }
}

@media (min-width: 1056px) {
  .tds-modal-xs {
    width: 31.25%
  }

  .tds-modal-sm {
    width: 43.75%
  }

  .tds-modal-md {
    width: 62.5%
  }

  .tds-modal-lg {
    width: 75%
  }
}

@media (min-width: 1312px) {
  .tds-modal-xs {
    width: 31.25%
  }

  .tds-modal-sm {
    width: 37.5%
  }

  .tds-modal-md {
    width: 62.5%
  }

  .tds-modal-lg {
    width: 75%
  }
}

@media (min-width: 1584px) {
  .tds-modal-xs {
    width: 25%
  }

  .tds-modal-sm {
    width: 37.5%
  }

  .tds-modal-md {
    width: 50%
  }

  .tds-modal-lg {
    width: 75%
  }
}

@media (max-width: 320px) {

  .tds-modal-md,
  .tds-modal-lg,
  .tds-modal-sm {
    height: 100%
  }

  .tds-modal-md slot[name=actions]::slotted(*),
  .tds-modal-lg slot[name=actions]::slotted(*),
  .tds-modal-sm slot[name=actions]::slotted(*) {
    display: flex
  }
}

.header {
  display: flex;
  padding-bottom: var(--tds-spacing-element-8);
  position: sticky;
  top: 0;
  background-color: var(--tds-modal-background);
  z-index: 1
}

.header::before {
  content: " ";
  position: absolute;
  height: 16px;
  top: -16px;
  width: 100%;
  background-color: var(--tds-modal-background)
}

.header,
slot[name=header]::slotted(*) {
  color: var(--tds-modal-text);
  font: var(--tds-headline-05);
  letter-spacing: var(--tds-headline-05-ls);
  margin: 0;
  flex: 1
}

.body {
  color: var(--tds-modal-text);
  font: var(--tds-body-01);
  letter-spacing: var(--tds-body-01-ls);
  padding-bottom: 40px;
  overflow-y: visible;
  padding-right: 16px;
}

.body::-webkit-scrollbar {
  width: 5px;
  background-color: var(--tds-grey-300)
}

.body::-webkit-scrollbar-track {
  background-color: var(--tds-grey-300)
}

.body::-webkit-scrollbar-thumb {
  background-color: var(--tds-grey-500)
}

@media (min-width: 1056px) {
  .body {
    padding-right: 64px
  }
}

.tds-modal-backdrop {
  box-sizing: border-box;
  left: 0;
  top: 0;
  position: fixed;
  height: 100%;
  width: 100%;
  z-index: 700;
  background-color: var(--tds-modal-backdrop);
  padding: 0 16px
}

.tds-modal-backdrop * {
  box-sizing: border-box
}

@media (max-width: 320px) {
  .tds-modal-backdrop {
    padding: 0
  }
}

.tds-modal-backdrop.show {
  display: flex
}

.tds-modal-backdrop.hide {
  display: none
}

button.tds-modal-close {
  margin: 0 0 auto auto;
  background-color: transparent;
  border: 0;
  padding: 0;
  appearance: unset
}

.tds-modal-close {
  display: inline-block;
  height: auto;
  color: var(--tds-modal-icon);
  cursor: pointer
}

.tds-modal-close:focus {
  outline: 2px solid var(--tds-blue-400);
  outline-offset: -2px
}

@media (min-width: 320px) {
  .tds-modal-close {
    margin-left: var(--tds-spacing-element-16)
  }
}

@media (min-width: 1056px) {
  .tds-modal-close {
    margin-left: var(--tds-spacing-element-48)
  }
}

.tds-modal-close-btn {
  display: inline-block;
  height: auto;
  background-repeat: no-repeat;
  cursor: pointer
}

@media (min-width: 320px) {
  .tds-modal-close-btn {
    margin-left: var(--tds-spacing-element-16)
  }
}

@media (min-width: 1056px) {
  .tds-modal-close-btn {
    margin-left: var(--tds-spacing-element-48)
  }
}

.tds-modal-close-btn svg {
  fill: var(--tds-modal-icon)
}

.tds-modal-overflow {
  overflow: hidden
}

:host {
  left: 0;
  top: 0;
  position: fixed;
  height: 100%;
  width: 100%;
  z-index: 700;
  background-color: var(--tds-modal-backdrop);
  padding: 0 16px
}

@media (max-width: 320px) {
  :host {
    padding: 0
  }
}

:host.show {
  display: flex
}

:host.hide {
  display: none
}

:host .tds-modal-close {
  border: none;
  background-color: transparent
}

:host .tds-modal-close-btn {
  border: none;
  background-color: transparent
}

:host(.show) {
  display: flex
}

:host(.hide) {
  display: none
}
```

**Comment by mJarsater** [comment]
> e used them 😅
> 
> If that sounds too risky then we could instead just move the CSS with the `:host` class and leave the `import` as is for now and change that in 2.0. Then maybe a temporary separate .scss file can be created to hold the `:host` styling, and that is what the component will use.

Yeah, what we have said earlier in the team is that breaking changes only apply to what we document. So would not consider this a breaking change. Great job!

**Comment by alexandereneroth** [comment]
Great suggestion to analyze the dist stylesheet. Could be part of a future greater effort on testing. 
Created an issue here: https://tegel.atlassian.net/browse/CDEP-2721.
